### PR TITLE
Initial commit of System.Diagnostics.Contracts facade

### PR DESCRIPTION
**Pull request by FiveTimesTheFun**
I'll let Stephen's change go first and merge his modification to System.Diagnostics.Contracts.Tests.csproj in to mine.

### CODE REVIEW

**Comment by mellinoe** [comment]
Isn't this assembly a full facade? Do we need to build this with the tests for some reason?

**Comment by FiveTimesTheFun** [comment]
It is.  I did the same "reference the contract from the package and run against the assembly built from the repro" for consistency and to validate the facade.

**Comment by mellinoe** [comment]
I think we should just stick with testing against the facade from the NuGet package, until we have a plan for building the contracts/facades from here. Although this is similar, it's not exactly the same as what we do internally to build facades, as this is going through the partial facade code path in GenFacades. Thoughts, @weshaggard  ?

**Comment by weshaggard** [comment]
I agree with @mellinoe we should wait until we add support for facades and all the facades should come out together. I also don't know if there is any real value in adding the project reference to the facades as it doesn't actually contain any code. 

@FiveTimesTheFun I would prefer you actually build out the fall facade infrastructure as opposed to doing this one off project.

File: src/System.Diagnostics.Contracts/src/System.Diagnostics.Contracts.csproj [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+<Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
+    <Platform Condition="'$(Platform)' == ''">AnyCPU</Platform>
+    <AssemblyName>System.Diagnostics.Contracts</AssemblyName>
+    <OutputType>Library</OutputType>
+    <ProjectGuid>{27328738-30D1-434a-8867-5C8ACACB91A8}</ProjectGuid>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
```
Comment (weshaggard) [inline src/System.Diagnostics.Contracts/src/System.Diagnostics.Contracts.csproj]: For full facades I think we should have a unique property IsFullFacade

File: src/System.Diagnostics.Contracts/src/System.Diagnostics.Contracts.csproj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    <NuGetTargetFrameworkMoniker>ASP.NetCore, version=v5.0</NuGetTargetFrameworkMoniker>    
+  </PropertyGroup>
```
Comment (weshaggard) [inline src/System.Diagnostics.Contracts/src/System.Diagnostics.Contracts.csproj]: Where is your reference to mscorlib? I would still like to have the list of references (aka seed assemblies) explicitly add to the project so it is clear what the seeds are.

File: src/System.Diagnostics.Contracts/tests/System.Diagnostics.Contracts.Tests.csproj [hunk 3]
```diff
@@ -16,6 +16,14 @@
     <None Include="project.json" />
   </ItemGroup>
   <ItemGroup>
+    <!-- Compile tests against the contract, but copy our local-built implementation for testing -->  
+    <ProjectReference Include="..\src\System.Diagnostics.Contracts.csproj">
```
Comment (weshaggard) [inline src/System.Diagnostics.Contracts/tests/System.Diagnostics.Contracts.Tests.csproj]: As I called out in the comments I don't think this adds any real value so I wouldn't necessary do this for full facades. Also given that right now all the hacking in the metadata required for full facades. If we do end up needing project references as much of this should be handled by the facade projects themselves and not in the people referencing them.

**Comment by FiveTimesTheFun** [comment]
Thanks Wes and Eric - I'll start looking at the full facade infrastructure soon.

**Pull request closed without merging**